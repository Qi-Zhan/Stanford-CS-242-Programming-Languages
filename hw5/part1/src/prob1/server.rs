/* DO NOT EDIT THIS FILE. */

/***************
 * NOTE:
 *   To implement `cart.rs`, you only need to understand
 *   the signature and specification of `internal_login`.
 *   You do not need to understand other parts.
 ***************/

/* USER_DB:
 *   - private database that stores (id, pw, acct_num) information of all users. */
static USER_DB: [(&'static str, &'static str, u32); 3] = [
    ("id1", "pw1", 123),
    ("id2", "pw2", 456),
    ("id3", "pw3", 789),
];

/* internal_login(id, pw):
 *   - return `Some(acct_num)`, if `(id, pw, acct_num)` is in `USER_DB` (i.e., login succeeds).
 *   - return `None`, if login fails. */
pub fn internal_login(id: String, pw: String) -> Option<u32> {
    let mut acct_num_opt: Option<u32> = None;
    for user in USER_DB.iter() {
        if id == user.0 && pw == user.1 {
            acct_num_opt = Some(user.2);
            break;
        }
    }
    acct_num_opt
}
