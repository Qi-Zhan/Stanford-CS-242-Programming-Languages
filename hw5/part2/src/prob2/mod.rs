/* DO NOT EDIT THIS FILE. */

pub mod msg;
pub mod server;
pub mod client;
