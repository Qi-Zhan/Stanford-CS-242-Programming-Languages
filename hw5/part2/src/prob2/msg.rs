/* DO NOT EDIT THIS FILE. */

/***************
 * NOTE:
 *   To implement `client.rs`, you only need to understand
 *   the definition of `Sig` and `Pkt`.
 *   You do not need to understand other parts.
 ***************/

use std::fmt;

/**************************
 * Defn of Signal, Packet *
 **************************/
pub enum Sig {
    Syn, SynAck, Ack, Close, CloseAck,
}

#[derive(Clone, PartialEq, Copy)]
pub struct Pkt {
    pub buf: char,
    pub id : u32,
}

/*************************
 * Ops on Signal, Packet *
 *************************/
impl fmt::Debug for Pkt {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "(buf: '{}', id: {})", self.buf, self.id)
    }
}

impl fmt::Debug for Sig {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let text: &str =
            match self {
                Sig::Syn      => "Sig::Syn",
                Sig::SynAck   => "Sig::SynAck",
                Sig::Ack      => "Sig::Ack",
                Sig::Close    => "Sig::Close",
                Sig::CloseAck => "Sig::CloseAck",
            };
        write!(f, "{}", text)
    }
}
