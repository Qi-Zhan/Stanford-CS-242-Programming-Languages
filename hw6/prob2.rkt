#lang racket
(print-as-expression #f)
(provide (all-defined-out))

;========;
; Import ;
;========;
; import `throw` and `try_except`.
(require "prob1.rkt")

;===========;
; Problem 2 ;
;===========;
; Task: Implement `eval` using `throw` and `try_except`.
; Note: You can define any other helper functions.

(define (eval e)
    (try_except
        (lambda() (eval_one e))
        (lambda (msg) (printf "except: ~a\n" msg))
    )
)


(define (eval_one e)
    (cond [(number? e) e]
          [else (let* ([s (first e)]
                       [n1 (eval_one (second e))]
                       [n2 (eval_one (third e))])
                       (cond [(equal? s "+") (+ n1 n2)]
                             [(equal? s "-") (- n1 n2)]
                             [(equal? s "*") (* n1 n2)]
                             [(equal? s "/") (if (equal? n2 0) (throw "DivError") (/ n1 n2))]
                             [else (throw "OpError")]
                       )
                 )])
)
