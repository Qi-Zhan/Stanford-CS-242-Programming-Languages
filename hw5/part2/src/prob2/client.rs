/***************
 * NOTE:
 *   You can use `Sig`, `Pkt`, ..., `internal_send_pkts`
 *   which appear in the below `use` declarations,
 *   simply by `Sig`, `Pkt`, ..., `internal_send_pkts` without any prefixes.
 ***************/

#[allow(unused_imports)]
use crate::prob2::msg::{Pkt, Sig};
#[allow(unused_imports)]
use crate::prob2::server::{internal_send_pkts, internal_send_sig, Server};

// TODO: Implement the typestate structs below for the TCP client. You are free to add any fields
// that you think will be helpful to the struct definitions.
// Below shows the signature of the methods needed across the TCP client state machine.
// Note that not all methods may be implemented for every typestate struct.
//
//   pub fn new() -> T
//   pub fn send_syn(self, _: &mut Server) -> Result<T,T>
//   pub fn send_ack(self, _: &mut Server) -> T
//   pub fn send_pkts(self, _: &mut Server, _: &Vec<Pkt>) -> T
//   pub fn send_close(self, _: &mut Server) -> Result<T,T>
//   pub fn ids_sent(&self) -> Vec<u32>
//
// Here T denotes a type. Note that each T can be a different type.
//===== BEGIN_CODE =====//
pub struct Client {}
pub struct Initial {
    sent_packet_ids: Vec<u32>,
}
pub struct Syned {
    sent_packet_ids: Vec<u32>,
}
pub struct SynAcked {
    sent_packet_ids: Vec<u32>,
}
pub struct Closed {
    sent_packet_ids: Vec<u32>,
}

impl Client {
    pub fn new() -> Initial {
        let sent_packet_ids = Vec::new();
        return Initial { sent_packet_ids };
    }
}

impl Initial {

    pub fn send_syn(self, server: &mut Server) -> Result<Syned,Initial> {
        let return_signal: Option<Sig> = internal_send_sig(server, Sig::Syn);
        return match return_signal {
            Some(_) => Ok(Syned { sent_packet_ids: self.sent_packet_ids }),
            None => Err(Initial { sent_packet_ids: self.sent_packet_ids }),
        }
    }

    pub fn ids_sent(&self) -> Vec<u32> {
        return self.sent_packet_ids;
    }
}

impl Syned {

    pub fn send_ack(self, server: &mut Server) -> SynAcked {
        internal_send_sig(server, Sig::Ack);
        return SynAcked { sent_packet_ids: self.sent_packet_ids }
    }

    pub fn ids_sent(&self) -> Vec<u32> {
        return self.sent_packet_ids;
    }

}

impl SynAcked {

    pub fn send_pkts(self, server: &mut Server, packets: &Vec<Pkt>) -> SynAcked {
        let new_sent_packet_ids: Vec<u32> = internal_send_pkts(server, packets);
        new_sent_packet_ids.extend(self.sent_packet_ids);
        return SynAcked { sent_packet_ids: new_sent_packet_ids}
    }

    pub fn send_close(self, server: &mut Server) -> Result<Closed, SynAcked> {
        let return_signal: Option<Sig> = internal_send_sig(server, Sig::Ack);
        match return_signal {
            Some(_) => return Ok(Closed{ sent_packet_ids: self.sent_packet_ids }),
            None => return Err(SynAcked{ sent_packet_ids: self.sent_packet_ids }),
        }
    }

    pub fn ids_sent(&self) -> Vec<u32> {
        return self.sent_packet_ids;
    }

}

impl Closed {

    pub fn send_ack(self, _: &mut Server) -> Initial {
        let sent_packet_ids = Vec::new();
        return Initial { sent_packet_ids}
    }

    pub fn ids_sent(&self) -> Vec<u32> {
        return self.sent_packet_ids;
    }

}

//===== END_CODE =====//
