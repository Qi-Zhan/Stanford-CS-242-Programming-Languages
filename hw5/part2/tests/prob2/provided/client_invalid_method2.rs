extern crate part2;
#[allow(unused_imports)]
use part2::prob2::msg::Pkt;
#[allow(unused_imports)]
use part2::prob2::server::{Server, RecvType};
#[allow(unused_imports)]
use part2::prob2::client::Client;

fn main() {
    let mut server: Server = Server::new(RecvType::All);
    let initial = Client::new();
    match initial.send_syn(&mut server) {
        Err(_syn_failed) =>
            assert!(false),
        Ok (syned) => {
            let synacked = syned.send_ack(&mut server);
            let synacked = synacked.send_pkts(&mut server, &vec![Pkt{buf:'a', id:1}]);
             // `send_ack` method should be inaccessible in synacked state.
            synacked.send_ack(&mut server);
        }
    }
}
