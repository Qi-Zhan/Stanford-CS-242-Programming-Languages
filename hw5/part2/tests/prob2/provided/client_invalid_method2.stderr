error[E0599]: no method named `send_ack` found for struct `SynAcked` in the current scope
  --> tests/prob2/provided/client_invalid_method2.rs:19:22
   |
19 |             synacked.send_ack(&mut server);
   |                      ^^^^^^^^ method not found in `SynAcked`
